pandas/_libs/tslib.pyx

```diff
@@ -429,17 +429,32 @@ def array_with_unit_to_datetime(
             fvalues = iresult.astype('f8') * m
             need_to_iterate = False
 
+        # GH20445
+        if values.dtype.kind == "f":
+            fresult = values.astype('f8', casting='same_kind', copy=False)
+            # fill by comparing to NPY_NAT constant
+            mask = fresult == NPY_NAT
+            fresult[mask] = 0.0
+            m_as_float = <float64_t> m
+            fvalues = fresult.astype('f8') * m_as_float
+            need_to_iterate = False
+
         # check the bounds
         if not need_to_iterate:
-
             if ((fvalues < Timestamp.min.value).any()
                     or (fvalues > Timestamp.max.value).any()):
                 raise OutOfBoundsDatetime(f"cannot convert input with unit '{unit}'")
-            result = (iresult * m).astype('M8[ns]')
-            iresult = result.view('i8')
-            iresult[mask] = NPY_NAT
-            return result, tz
-
+            # GH20445
+            if values.dtype.kind == 'i':
```

Comment:
Are you sure this branch do anything? Casting shouldn't impact performance and in this case conversion should be handled implicitly if you just use `m`

Reply:
For the current example I think that's right: I still get similar runtimes for ints and floats after cleaning this up.

That said, with the pared-down version I get into trouble on the timezone tests, so it seems like that was somehow being taken care of.
```
===================================================================================== short test summary info ======================================================================================
FAILED pandas/tests/tools/test_to_datetime.py::TestOrigin::test_epoch[us-timestamp] - AssertionError: numpy array are different
FAILED pandas/tests/tools/test_to_datetime.py::TestOrigin::test_epoch[us-pydatetime] - AssertionError: numpy array are different
FAILED pandas/tests/tools/test_to_datetime.py::TestOrigin::test_epoch[us-datetime64] - AssertionError: numpy array are different
FAILED pandas/tests/tools/test_to_datetime.py::TestOrigin::test_epoch[us-str_1960] - AssertionError: numpy array are different
```